# <h1>Homework 1

# <h2>Importing Libraries

In [1]:
import numpy as np
from scipy.optimize import minimize
from scipy.optimize import LinearConstraint
from scipy.optimize import Bounds


# <h2>Defining Problem

In [2]:
bounds = Bounds([-10,-10,-10,-10,-10],[10,10,10,10,10])
linear_constraints = LinearConstraint([[1,3,0,0,0],
                                       [0,0,1,1,-2],
                                       [0,1,0,0,-1]]
                                      ,[0,0,0],[0,0,0])

# <h2>Defining Cost Function

In [3]:
def cost_func(x):
    return (x[0]-x[1])**2+(x[1]+x[2]-2)**2+(x[3]-1)**2+(x[4]-1)**2

# <h2>Solving Problem
# <h4>Initial guess
$x_{0} = \begin{bmatrix}
1&2&3&4&5
\end{bmatrix}$
# <h4>Alternate guess
$x_{0} = \begin{bmatrix}
-5&-4&-3&-2&-1
\end{bmatrix}$

In [4]:
x0 = np.array([1,2,3,4,5])
sol = minimize(cost_func,x0,method='SLSQP',constraints=linear_constraints,bounds=bounds)


# <h2> Solution

In [5]:
print(sol.fun,'\n',sol.x)

4.09302325588931 
 [-0.76743912  0.25581304  0.62791188 -0.1162858   0.25581304]


The optimization of this cost function resulted in a minimum value of $4.0930$.<br/>
Solution with $x_{0} =
 \begin{bmatrix}
   1 & 2 & 3 & 4 & 5
  \end{bmatrix}
$ is <br/>$x_{sol} = \begin{bmatrix}
-0.76743912 & 0.25581304 & 0.62791188 & -0.1162858  & 0.25581304
\end{bmatrix}
$
<br/>
Updating the parameters to $x_{0} =
 \begin{bmatrix}
   -5 & -4 & -3 & -2 & -1
  \end{bmatrix}
$ generates the solution as $x_{sol} = \begin{bmatrix}
-0.76744227 & 0.25581409 & 0.62790622 & -0.11627804 & 0.25581409
\end{bmatrix}
$
<br/>
It can thus be observed that the change in optimum solution of $x$ with change of initial guess only changes
after the $4^{th}$ decimal place. In other words, we can say that there is no significant change in optimum
solution. <br/>
Thus, updating the initial guess **did not** change the optimum solution